### ISSUE TITLE
The usage of L_c

### ISSUE THREAD

**Issue by Ther-nullptr** [princeton-nlp/CoFiPruning#18]
![image](https://user-images.githubusercontent.com/71265304/181204350-991474d7-8e39-4b08-b422-26120a6df844.png)
I do not understand how this loss works ---- since $\lambda_1$ and $\lambda_2$ are 0 as default, I find that sometimes the loss maybe a negative number sometimes.

**Comment by xiamengzhou** [comment]
This part guarantees that a target sparsity is met after optimization by enforcing a lagrangian constraint. Removing this part would break the pruning process. Please refer to [Wang et al. 2021](https://arxiv.org/pdf/1910.04732.pdf) for more details

**Comment by Ther-nullptr** [comment]
get it!

**Comment by Ther-nullptr** [comment]
Btw, why use negative lr(`train.py`)?
```python
lagrangian_params = [{
                    "params": [p for n, p in self.l0_module.named_parameters() if "lambda" in n],
                    "weight_decay": 0.0,
                    "lr": -self.additional_args.reg_learning_rate
                }]
```

**Comment by xiamengzhou** [comment]
Solving the constrained optimization problem with Lagrange multipliers is a min-max optimization problem. You minimize the loss (including the task loss and the sparsity constraint) by optimizing over main mode parameters and l0 module's parameters and maximizing the loss over the Lagrange multipliers to enforce the constraint to be met. Since we need to maximize the loss for the Lagrange multipliers, we set its LR as a negative value.

**Comment by xiamengzhou** [comment]
Hi, I am closing this issue now. Feel free to reopen it if you have more questions :)
